src/js_types.hpp

```diff
@@ -80,18 +81,16 @@ struct Context {
 
 class TypeErrorException : public std::invalid_argument {
 public:
-    std::string const& prefix() const { return m_prefix; }
-    std::string const& type() const { return m_type; }
-
-    TypeErrorException(std::string prefix, std::string type) : 
-        std::invalid_argument(prefix + " must be of type: " + type),
-        m_prefix(std::move(prefix)),
-        m_type(std::move(type)) 
-        {}
-
-private:
-    std::string m_prefix;
-    std::string m_type;
+    TypeErrorException(StringData object_type, StringData property,
+                       std::string const& type, std::string const& value)
+    : std::invalid_argument(util::format("%1.%2 must be of type '%3', got (%4)",
```

Comment:
Why not "'%4'"?

Reply:
Results in some misleading messages when the value is an array (looks too much like the value provided was just a string).